branches/mono-merge/mono-merge.8
.\" Copyright 1996-2003 Jeremy Messenger <mezz7@cox.net>.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgment:
.\"      This product includes software developed by John D. Polstra.
.\" 4. The name of the author may not be used to endorse or promote products
.\"    derived from this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
.\" THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\" $Id: mono-merge.8,v 1.3 2005/05/15 07:06:40 tmclau02 Exp $
.\"
.Dd September 04, 2004
.Os FreeBSD
.Dt MONO-MERGE 8
.Sh NAME
.Nm mono-merge
.Nd tool to merge BSD# CVS repository with FreeBSD ports tree
.Sh SYNOPSIS
.Nm
.Op Fl aDlpStuvx
.Op Fl c Ar cvsroot
.Op Fl d Ar directory
.Op Fl m Ar module
.Op Fl s Ar directory
.Sh DESCRIPTION
The
.Nm
script automates the procedure of merging the latest versions of Mono
related ports from the BSD# CVS repository with your current ports
collection. The script can also check out the BSD# ports from CVS.
.Pp
The ports from the BSD# repository are development versions of ports in
the FreeBSD ports tree or Mono/C# ports not currently available there.
The ports in this tree are meant to be a testing ground for changes to
Mono and related packages, for the introduction of new applications, or
for software with no applicable use on FreeBSD currently (like .NET data
providers not currently used by any ports). They are generally
experimental versions and are not in the FreeBSD tree precisely because
they are experimental and possibly unstable. Users are expected to help
by reporting and debugging any problems found when building and using
these ports to the BSD# project
.Aq bsd-sharp-list@forge.novell.com  .
.Sh OPTIONS
The following options are supported by
.Nm :
.Bl -tag -width Fl
.It Fl a
Same as
.Fl u
.Fl l
.Fl p .
By default,
.Nm
will simply merge the existing BSD# ports tree with the existing FreeBSD
ports tree. This option will update both trees first, and tell you which
programs will need to be upgraded.
.It Fl c
Specifies the CVSROOT, overriding any
.Cm CVSROOT
specification in
.Ar /usr/local/etc/mono-merge.cfg .
The default
.Cm CVSROOT
is
.Ar anonymous@forgecvs1.novell.com:/cvsroot/bsd-sharp .
.It Fl D
Causes
.Nm
to reset lang/mono port dependencies to
.Ar lang/mono-devel .
This is useful when switching to using the Mono development branch port. When switching to the
.Ar lang/mono-devel
port for the first time the user should force a rebuild of all ports that were dependednt on
.Ar lang/mono
(and are now dependant on
.Ar lang/mono-devel )
to ensure that ports are built against the latest Mono installed.
.It Fl d
Specifies the path of the official ports, overriding any
.Cm DESTDIR
specification in
.Ar mono-merge.cfg .
The default path is
.Ar /usr/ports .
.It Fl l
Lists the ports for which the BSD# ports tree has provided a newer
version than existed in the FreeBSD ports tree; overridding any
.Cm pkgversion
specification in
.Ar mono-merge.cfg .
The default is not to list the ports needing an update.
.It Fl m
Specifies the CVS module to checkout from BSD# CVS. In general, this
should always be
.Ar FreeBSD-ports .
.Pp
This option overrides any
.Cm MODULE
specification in
.Ar mono-merge.cfg .
The default CVS module is
.Ar ports .
.It Fl p
Updates the official ports collection and checks out the BSD# ports.
You should fill out the
.Cm SUPFILE
in
.Nm
before using this option. The default is not to cvsup the latest
official ports tree.
.It Fl S
Causes
.Nm
to reset
.Ar lang/mono
port dependencies to
.Ar lang/mono-svn .
This is useful when switching to using the Mono SVN snapshot port. When switching to the
.Ar lang/mono-svn
port for the first time the user should force a rebuild of all ports that were dependednt on
.Ar lang/mono
(and are now dependant on
.Ar lang/mono-svn )
to ensure that ports are built against the latest Mono installed.
.It Fl s
Specifies the path to the BSD# ports without the trailing module name
(e.g. without the trailing /ports), overriding any
.Cm SRCDIR
specification in
.Ar mono-merge.cfg .
There is no default path, so you must either specify this option, or
edit
.Ar mmono-merge.cfg
and fill in a value for
.Cm SRCDIR .
For example, if you keep the BSD# ports tree in
/usr/ports/bsd-sharp/ports, specify "/usr/ports/bsd-sharp" as the value
of
.Cm SRCDIR
or the argument to
.Fl s .
.It Fl u
Updates the local copy of the BSD# ports repository before merging,
overridding any
.Cm updating
specification in
.Ar mono-merge.cfg .
The default is not to update the local copy of the BSD# repository
before merging.
.It Fl v
Enables verbose output, overridding any
.Cm VERBOSE
specification in
.Ar mono-merge.cfg .
The default is not to display verbose output.
.It Fl x
This option merges the experimental versions of applications after the
usual development versions have been merged.
.Pp
The BSD# repository is a testbed for Mono and related applications that
will eventually make their way into the official FreeBSD ports tree.
.Sh VARIABLES
You can add variable definitions to
.Ar /usr/local/etc/mono-merge.cfg .
The
.Ar mono-merge.cfg
file will override any varibles in
.Nm ,
so you can copy the varibles from
.Nm
into
.Ar mono-merge.cfg .
By using
.Ar mono-merge.cfg ,
you do not have to specify any command line arguments to
.Nm .
.Pp
You can add the following variables to
.Ar mono-merge.cfg :
.Bl -tag -width Cm
.It Cm CVSUP_CMD= Ns Ar command
Set the command used to update the official ports collection. This is
only used if
.Cm update_main
is set to
.Ar yes
or if the
.Fl p
option is specified. Before using this option, you must set the
.Cm SUPFILE
variable. The default command is
.Ar /usr/local/bin/cvsup -g .
.It Cm DESTDIR= Ns Ar path
Set the path to the local FreeBSD ports tree. This id overridden by the
.Fl d
option. The default path is
.Ar /usr/ports .
.It Cm MODULE= Ns Ar module
Specify which BSD# CVS module will be checked out. In general, this
should always be left to the default,
.Ar ports .
This variable is overridden by the
.Fl m
option. The default CVS module is
.Ar ports .
.It Cm PKGVERSION_ARGS= Ns Ar options
Set the options for
.Cm PKGVERSION_CMD
to
.Ar options .
This is only useful if
.Cm pkgversion
is set to
.Ar yes
or the
.Fl l
option is specified. The default options are
.Ar -l \< .
.It Cm PKGVERSION_CMD= Ns Ar command
Specify the command used to report port versions.
This is only useful if
.Cm pkgversion
is set to
.Ar yes
or the
.Fl l
option is specified. The default command is
.Ar /usr/sbin/pkg_version .
.It Cm SRCDIR= Ns Ar path
Set the path to the local copy of the BSD# repository to
.Ar path
without the trailing module name (e.g. without the trailing /ports).
This variable is overridden by the
.Fl s
options. There is no default
path, so you must either specify this variable or set the
.Fl s
option.
.It Cm SUPFILE= Ns Ar path
Set the path to your ports supfile to
.Ar path .
The default is to auto-detect this.
.It Cm VERBOSE= Ns Ar yes|no
Set to
.Ar yes
to enable verbose output.  This is overridden by the
.Fl v
option. The default is
.Ar no .
.It Cm pkgversion= Ns Ar yes|no
Set to
.Ar yes
to show a list of outdated ports after
.Nm
finishes the merge.  This is overridden by the
.Fl l
option. The default is
.Ar no .
.It Cm update_main= Ns Ar yes|no
Set to
.Ar yes
to update the official ports tree before merging.  This is overridden by the
.Fl p
option. You should also make sure the
.Cm SUPFILE
variable is correct before use this option. The default is
.Ar no .
.It Cm updating= Ns Ar yes|no
Set to
.Ar yes
to update the local copy of the BSD# repository before merging. This is
overridden by the
.Fl u
option. The default is
.Ar no .
.El
.Sh FILES
.Bl -tag -width Pa
.It Pa $ Ns Ev PATH Ns Pa /mono-merge
Path to the script itself.
.It Pa /usr/local/etc/mono-merge.cfg
See the
.Cm VARIABLES
section above to learn more about this file.
.El
.Sh AUTHORS
.An Joe Marcus Clarke Aq marcus@FreeBSD.org
.An Martin Klaffenboeck Aq martin.klaffenboeck@gmx.at
.An Franz Klammer Aq klammer@webonaut.com
.An Adam Weinberger Aq adamw@FreeBSD.org
.An Tom McLaughlin (BSD# specific changes) Aq tmclaugh@sdf.lonestar.org
.Pp
.An This manual page was written by Jeremy Messenger Aq mezz7@cox.net .
.An The Cm DESCRIPTION No was written by Khairil Yusof Aq kaeru@pd.jaring.my .
.Sh LICENCE AND DISCLAIMER
BSD License


branches/mono-merge/mono-merge.sh
#!/bin/sh
#-
# Copyright (c) 2002-2004 FreeBSD GNOME Team <freebsd-gnome@FreeBSD.org>
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
# SUCH DAMAGE.
#
# $Id: mono-merge.sh,v 1.11 2005/05/31 01:40:26 mezz7 Exp $
#

# These variables are overrideable in /usr/local/etc/${SCRIPTNAME}.cfg.

SRCDIR=${MARCUSMERGE_SRCDIR:=""}
			# The path to the BSD# ports without the trailing
			# module name (i.e. without the trailing /FreeBSD-ports)
DESTDIR=${MARCUSMERGE_DESTDIR:="/usr/ports"}
			# The path to the official ports collection.
VERBOSE=${MARCUSMERGE_VERBOSE:="no"}
			# If you want verbose output.
SUPFILE=${MARCUSMERGE_SUPFILE:=""}
			# The path your ports-supfile (default: auto).
TARBALL=${MARCUSMERGE_TARBALL:="no"}
			# Set to yes to download a tarball instead of using CVS.
MODULE=${MARCUSMERGE_PORTS:="FreeBSD-ports"}
			# CVS module to checkout.

#For Novell repo:
#CHECKOUT_CMD=${CHECKOUT_CMD:="cvs -z3 -d:ext:anonymous@forgecvs1.novell.com:/cvsroot/bsd-sharp checkout -P "}
#UPDATE_CMD=${UPDATE_CMD:="cvs -z3 update -Pd -A "}

#For google repo:
CHECKOUT_CMD=${CHECKOUT_CMD:="svn checkout http://bsd-sharp.googlecode.com/svn/trunk  "}
UPDATE_CMD=${UPDATE_CMD:="svn update "}

# You do not have to change anything beyond this line.

PKGVERSION_CMD="/usr/sbin/pkg_version"
PKGVERSION_ARGS="-l \<"

CVSUP_CMD="/usr/local/bin/cvsup -g"
FETCH_CMD="/usr/bin/fetch"
FETCH_ARGS="-4 -q"

update_main="no"
pkgversion="no"
updating="no"
show_updating="no"
use_monodev="no"
use_monosvn="no"

# End list of overrideable variables

SCRIPTNAME=`basename $0`

if [ -f /usr/local/etc/${SCRIPTNAME}.cfg ]; then
.    /usr/local/etc/${SCRIPTNAME}.cfg
elif [ -f /usr/local/etc/`echo ${SCRIPTNAME} | sed -e 's/.sh$//'`.cfg ]; then
.    /usr/local/etc/`echo ${SCRIPTNAME} | sed -e 's/.sh$//'`.cfg
fi

SCRIPTNAME=`basename $0`

CVSROOT=":ext:anonymous@forgecvs1.novell.com:/cvsroot/bsd-sharp"
TARBALL_URL="http://forge.novell.com/modules/xfmod/cvs/cvsbrowse.php/bsd-sharp/${MODULE}"
RMPORTS="RMPORTS"
MCVER_FILE="MCVER"

MCVER="2008021701"

get_tmpfile()
{
    template=$1
    tmpfile=""

    if [ -n "${MC_TMPDIR}" -a -d "${MC_TMPDIR}" ]; then
	tmpfile="${MC_TMPDIR}/${template}.XXXXXX"
    elif [ -n "${TMPDIR}" -a -d "${TMPDIR}" ]; then
	tmpfile="${TMPDIR}/${template}.XXXXXX"
    elif [ -d "/var/tmp" ]; then
	tmpfile="/var/tmp/${template}.XXXXXX"
    elif [ -d "/tmp" ]; then
	tmpfile="/tmp/${template}.XXXXXX"
    elif [ -d "/usr/tmp" ]; then
	tmpfile="/usr/tmp/${template}.XXXXXX"
    else
	return 1
    fi

    tmpfile=`mktemp -q ${tmpfile}`

    echo ${tmpfile}

    return 0
}

download_tarball()
{
    tmpfile=`get_tmpfile mctmp`

    if [ $? != 0 ]; then
	printf "Unable to create a temporary file for the downloaded tarball.\nPlease set your MC_TMPDIR environment variable to a suitable temporary directory,\nand run this script again.\n"
	exit 1
    fi

    echo "===> Downloading the tarball for the ${MODULE} module."
    ${FETCH_CMD} ${FETCH_ARGS} -o ${tmpfile} "${TARBALL_URL}/${MODULE}.tar.gz?tarball=1"
    if [ $? != 0 ]; then
        printf "Failed to download ${MODULE}.tar.gz from ${TARBALL_URL}.\n\n"
        exit 1
    fi
    echo "===> Download done."
    echo "===> Extracting ${tmpfile} to ${SRCDIR}."
    mkdir -p ${SRCDIR}
    tar -C ${SRCDIR} -xvjf ${tmpfile}
    echo "===> Extraction done."
    rm -f ${tmpfile}
}

usage="usage: ${SCRIPTNAME} [-s <directory>] [-d <directory>] [-c <checkout command>] [-P <update command>] [-m <module>] [-t] [-u] [-v] [-l] [-p] [-a] [-x] [-U] [-D|-S]"

while getopts " hvulpaxUDSs:d:c:m:P:" i
do
    case "$i" in
	s)
		SRCDIR="$OPTARG";
		continue;;
	d)
		DESTDIR="$OPTARG";
		continue;;
	c)
		CHECKOUT_CMD="$OPTARG";
		continue;;
	m)
		MODULE="$OPTARG";
		continue;;
	P)
		UPDATE_CMD="$OPTARG";
		continue;;
	v)
		VERBOSE="yes";
		continue;;
	t)
		TARBALL="yes";
		continue;;
	u)
		updating="yes";
		continue;;
	l)
		pkgversion="yes";
		continue;;
	p)
		update_main="yes";
		continue;;
	a)
		updating="yes";
		pkgversion="yes";
		update_main="yes";
		continue;;
	x)
		experimental="yes";
		continue;;
	U)
		show_updating="yes";
		continue;;
	D)
		use_monodev="yes";
		continue;;
	S)
		use_monosvn="yes";
		continue;;
	h)
		echo $usage;
		exit 1;;
	*)
		echo $usage;
		exit 1;;
    esac
done

if [ -z "${SRCDIR}" ]; then
    echo "You did not specify a SRCDIR; aborting..."
    exit 1
fi

if [ -z "${MODULE}" ]; then
    echo "You did not specify a MODULE; aborting..."
    exit 1
fi

if [ ${use_monodev} = "yes" -a ${use_monosvn} = "yes" ]; then
    echo "Please select mono-devel or mono-svn but not both."
    exit 1
fi

# This function does, well, everything. It takes the module name as an argument.
run_marcusmerge() {
moduledir="$1"
if [ ${SRCDIR} = ${DESTDIR} -o "${SRCDIR}/${moduledir}" = ${DESTDIR} ]; then
    printf "Your SRCDIR is set to the same path as your DESTDIR.\nPlease set it to a path where we can store the BSD# ports.\n\n"
    exit 1
fi

first_time=0

#if [ "${TARBALL}" != "yes" ]; then
#    printf "First we have to check out the ${moduledir} module from\nhttp://forge.novell.com/modules/xfmod/project/?bsd-sharp\n\nCVS Password: anonymous\n\n"
#fi

# If the SRCDIR doesn't exist, we need to download the repo.  Currently, we
# can do this via CVS or from a tarball generated periodically.
if [ ! -d ${SRCDIR} -o ! -d ${SRCDIR}/${moduledir} ]; then
    first_time=1
    if [ "${TARBALL}" = "yes" ]; then
        download_tarball
    else
        echo "===> Checking out the ${moduledir} module."
        mkdir -p ${SRCDIR}
		if [ $? != 0 ]; then
			printf "Failed to create directory ${SRCDIR}.\n\n"
			exit 1
		fi
        cd ${SRCDIR}
		${CHECKOUT_CMD}${moduledir}
		if [ $? != 0 ]; then
		    printf "Failed to check out the BSD# ports tree.\n\n"
		    exit 1
		fi
	        echo "===> Checkout done."
	    fi
fi

# Update the BSD# repo if so requested. Don't update if we just checked
# out (or downloaded) the repo for this first time.
if [ ${updating} = "yes" -a ${first_time} = 0 ]; then
    echo "===> Updating the BSD# ports tree"
    if [ "${TARBALL}" = "yes" ]; then
	rm -rf "${SRCDIR}/${moduledir}"
        download_tarball
    else
        cd "${SRCDIR}/${moduledir}"
        ${UPDATE_CMD}
		if [ $? != 0 ]; then
	    	printf "Failed to update the BSD# ports tree.\n\n"
		    exit 1
		fi
    fi
    echo "===> Updating done."
fi

# Determine if this repo needs a newer mono-merge.
mcver="0"
if [ -f "${SRCDIR}/${moduledir}/${MCVER_FILE}" ]; then
    mcver=`cat "${SRCDIR}/${moduledir}/${MCVER_FILE}"`
    if [ "${mcver}" != "${MCVER}" ]; then
        echo "This repository requires a newer version of mono-merge.  Please go to http://code.google.com/p/bsd-sharp/downloads/list to download the latest version." | fmt 75 79
        exit 1
    fi
fi

# Handle ports that may have been moved or deleted in the BSD# repo.
if [ -f "${SRCDIR}/${moduledir}/${RMPORTS}" ]; then
    echo "===> Removing obsolete ports from main tree"
    for i in `cat ${SRCDIR}/${moduledir}/${RMPORTS} | cut -d '|' -f1`; do
	if [ "${VERBOSE}" = "yes" ]; then
	    echo " Removing obsolete port ${i}"
	fi
	rm -rf ${DESTDIR}/${i}
    done
    if [ "${VERBOSE}" = "yes" ]; then
	echo "Adding removed ports to MOVED"
    fi
    date=`date "+%Y-%m-%d"`
    cat ${SRCDIR}/${moduledir}/${RMPORTS} | sed -e "s|%%DATE%%|${date}|g" \
    	>> ${DESTDIR}/MOVED
    echo "===> done."
fi

echo "===> Merging files into the ports directory"
echo "${SRCDIR}/${moduledir} --> ${DESTDIR}"

if [ -d ${SRCDIR}/${moduledir}/Mk ]; then
    if [ "${VERBOSE}" = "yes" ]; then
	echo "Merging Mk files"
    fi
    if [ ! -d ${DESTDIR}/Mk ]; then
	if [ "${VERBOSE}" = "yes" ]; then
	    echo "Creating nonexistent destination category directory: Mk"
	fi
	mkdir -p ${DESTDIR}/Mk
    fi
    for mk in `ls -1 ${SRCDIR}/${moduledir}/Mk/*.mk`; do
	cp ${mk} ${DESTDIR}/Mk
    done
fi

for category in `ls -1 ${SRCDIR}/${moduledir}`; do
    if [ ${category} = "CVS" ]; then
	continue
    fi
    for port in `ls -1 ${SRCDIR}/${moduledir}/${category}`; do
	if [ ${port} = "CVS" ]; then
	    continue
	fi
	if [ ! -f ${SRCDIR}/${moduledir}/${category}/${port}/Makefile ]; then
	    continue
	fi
	if [ ! -d ${DESTDIR}/${category} ]; then
	    if [ "${VERBOSE}" = "yes" ]; then
		echo "Creating nonexistent destination category directory: ${category}"
            fi
	    mkdir -p ${DESTDIR}/${category}
	fi
	if [ -d ${DESTDIR}/${category}/${port} ]; then
	    if [ "${VERBOSE}" = "yes" ]; then
		echo "Removing old directory for: ${category}/${port}"
	    fi
	    rm -rf ${DESTDIR}/${category}/${port}
	fi

	cd ${SRCDIR}/${moduledir}/${category}
	tar --exclude *CVS* -cf - ${port} | \
		tar -xf - -C ${DESTDIR}/${category}
    done
done
}	# end of run_marcusmerge()

# Try to get our cvsup supfile if we will be updating the main ports tree.
if [ ${update_main} = "yes" ]; then
    if [ -z "${SUPFILE}" ]; then
	SUPFILE=`cd ${DESTDIR} && make -V PORTSSUPFILE`
    fi
    if [ -z "${SUPFILE}" ]; then
        printf "Update of main ports tree requested but you forgot\nto set SUPFILE in this script; aborting...\n\n"
        exit 1
    fi
    if [ ! -f ${SUPFILE} ]; then
        printf "Update of main ports tree requested but I could not\nfind SUPFILE: ${SUPFILE}; aborting...\n\n"
        exit 1
    fi
fi

# Update the main ports tree if so requested.
if [ ${update_main} = "yes" ]; then
    echo "===> Updating the main ports tree"
    if [ ${VERBOSE} = "yes" ]; then
        ${CVSUP_CMD} ${SUPFILE}
    else
        ${CVSUP_CMD} ${SUPFILE} -L0
    fi
    if [ $? != 0 ]; then
	printf "Failed to cvsup the ports tree.\n\n"
	exit 1
    fi
    echo "===> Updating done."
fi

run_marcusmerge ${MODULE}
if [ "${experimental}" = "yes" ]; then
	run_marcusmerge ${EXPMODULE}
fi

echo "===> Merging done."; echo

if [ ${use_monodev} = "yes" ]; then
    if [ ! -f "${SRCDIR}/${MODULE}/lang/mono-devel/Makefile" ]; then
        echo "===> WARNING: mono-devel doesn't exist, fallback to mono"; echo
        use_monodev="no"
    fi
fi

if [ ${use_monosvn} = "yes" ]; then
    if [ ! -f "${SRCDIR}/${MODULE}/lang/mono-svn/Makefile" ]; then
        echo "===> WARNING: mono-svn doesn't exist, fallback to mono"; echo
        use_monosvn"no"
    fi
fi

if [ ${use_monodev} = "yes" ]; then
    for _port in `cat ${SRCDIR}/${MODULE}/PORTSLIST`; do
        if [ -f ${DESTDIR}/${_port}/Makefile ]; then
            sed -i.bak 's|${PORTSDIR}/lang/mono$|${PORTSDIR}/lang/mono-devel|g' \
                ${DESTDIR}/${_port}/Makefile
	    sed -i.bak 's|${PORTSDIR}/lang/mono \$|${PORTSDIR}/lang/mono-devel|g' \
	    	${DESTDIR}/${_port}/Makefile
            sed -i.bak 's|${PORTSDIR}/lang/mono-svn$|${PORTSDIR}/lang/mono-devel|g' \
                ${DESTDIR}/${_port}/Makefile
	    sed -i.bak 's|${PORTSDIR}/lang/mono-svn$|${PORTSDIR}/lang/mono-devel|g' \
	    	${DESTDIR}/${_port}/Makefile
            rm ${DESTDIR}/${_port}/Makefile.bak
        fi
    done
    echo "===> setting mono dependency to lang/mono-devel done."; echo

elif [ ${use_monosvn} = "yes" ]; then
    for _port in `cat ${SRCDIR}/${MODULE}/PORTSLIST`; do
        if [ -f ${DESTDIR}/${_port}/Makefile ]; then
	    sed -i.bak 's|${PORTSDIR}/lang/mono$|${PORTSDIR}/lang/mono-svn|g' \
	        ${DESTDIR}/${_port}/Makefile
	    sed -i.bak 's|${PORTSDIR}/lang/mono \$|${PORTSDIR}/lang/mono-svn|g' \
	    	${DESTDIR}/${_port}/Makefile
	    sed -i.bak 's|${PORTSDIR}/lang/mono-devel$|${PORTSDIR}/lang/mono-svn|g' \
	        ${DESTDIR}/${_port}/Makefile
	    sed -i.bak 's|${PORTSDIR}/lang/mono-devel \$|${PORTSDIR}/lang/mono-svn|g' \
	    	${DESTDIR}/${_port}/Makefile
	    rm ${DESTDIR}/${_port}/Makefile.bak
        fi
    done
    echo "===> setting mono dependency to lang/mono-svn done."; echo

else
    for _port in `cat ${SRCDIR}/${MODULE}/PORTSLIST`; do
        if [ -f ${DESTDIR}/${_port}/Makefile ]; then
            sed -i.bak 's|${PORTSDIR}/lang/mono-svn$|${PORTSDIR}/lang/mono|g' \
                ${DESTDIR}/${_port}/Makefile
            sed -i.bak 's|${PORTSDIR}/lang/mono-devel$|${PORTSDIR}/lang/mono|g' \
                ${DESTDIR}/${_port}/Makefile
            rm ${DESTDIR}/${_port}/Makefile.bak
        fi
    done
    echo "===> setting mono dependency to lang/mono done."; echo
fi

if [ "${pkgversion}" = "yes" ]; then
    echo "===> Listing outdated ports..."
    ${PKGVERSION_CMD} ${PKGVERSION_ARGS}
    echo "===> Listing done."
fi

echo ""

if [ "${show_updating}" = "yes" ]; then
    less ${SRCDIR}/${MODULE}/UPDATING
fi

echo ""

echo "Merge finished, please follow your regular updating procedures" | /usr/bin/fmt 75 79
